applets/clock/clock.c

```diff
@@ -888,105 +893,181 @@ create_calendar (ClockData *cd)
+                                                                      &x, &y);
+
+                /*Get the panel dimensions*/
+                toplevel = gtk_widget_get_toplevel (cd->applet);
+                panelwin = gtk_widget_get_window (toplevel);
+                gdk_window_get_geometry (panelwin, NULL, NULL, &panel_w, &panel_h);
+
+                /*Set up GTK Layer Shell*/
+                gtk_layer_init_for_window (window);
+                gtk_layer_set_layer (window, GTK_LAYER_SHELL_LAYER_TOP);
+
+                switch (cd->orient) {
+                case MATE_PANEL_APPLET_ORIENT_RIGHT:
+                        gtk_layer_set_anchor (window, GTK_LAYER_SHELL_EDGE_LEFT, TRUE);
+                        if (y < (panel_h -h))
+                        {
+                                gtk_layer_set_anchor (window, GTK_LAYER_SHELL_EDGE_TOP, TRUE);
+                                gtk_layer_set_margin (window, GTK_LAYER_SHELL_EDGE_TOP, y);
+                        }
+                        else
+                                gtk_layer_set_anchor (window, GTK_LAYER_SHELL_EDGE_BOTTOM, TRUE);
 
-                gravity = GDK_GRAVITY_SOUTH_WEST;
+                        break;
+                case MATE_PANEL_APPLET_ORIENT_LEFT:
+                        gtk_layer_set_anchor (window, GTK_LAYER_SHELL_EDGE_RIGHT, TRUE);
+                        if (y < (panel_h - h))
+                        {
+                                gtk_layer_set_anchor (window, GTK_LAYER_SHELL_EDGE_TOP, TRUE);
+                                gtk_layer_set_margin (window, GTK_LAYER_SHELL_EDGE_TOP, y);
+                        }
+                        else
+                                gtk_layer_set_anchor (window, GTK_LAYER_SHELL_EDGE_BOTTOM, TRUE);
 
-                break;
-        }
+                        break;
+                case MATE_PANEL_APPLET_ORIENT_DOWN:
+                        gtk_layer_set_anchor (window, GTK_LAYER_SHELL_EDGE_TOP, TRUE);
+                        if (x < (panel_w -w))
+                        {
+                                gtk_layer_set_anchor (window, GTK_LAYER_SHELL_EDGE_LEFT, TRUE);
+                                gtk_layer_set_margin (window, GTK_LAYER_SHELL_EDGE_LEFT, x);
+                        }
+                        else
+                                gtk_layer_set_anchor(window, GTK_LAYER_SHELL_EDGE_RIGHT, TRUE);
+
+                        break;
+                case MATE_PANEL_APPLET_ORIENT_UP:
+                        gtk_layer_set_anchor (window, GTK_LAYER_SHELL_EDGE_BOTTOM, TRUE);
+                        if (x < (panel_w -w))
+                        {
+                                gtk_layer_set_anchor (window, GTK_LAYER_SHELL_EDGE_LEFT, TRUE);
+                                gtk_layer_set_margin (window, GTK_LAYER_SHELL_EDGE_LEFT, x);
+                        }
+                        else
+                                gtk_layer_set_anchor (window, GTK_LAYER_SHELL_EDGE_RIGHT, TRUE);
```

Comment:
Won't you have a problem similar to [the one in mate-media] when moving and re-orienting panels? (not tested though)

Reply:
I tested for that specifically and had no trouble, but will apply the fix from mate-media in case someone else gets different behavior. Not sure if the calendar window is made anew each time (the volume slider is just hidden and shown) or not

Reply:
Setting the unused anchors to FALSE for code durability/bug prevention reasons

Reply:
On further testing, I found that there was indeed a "stretching" issue-but only if the applet was moved or the panel orientation changed while the calendar is showing. No change I tried stopped it in once situation without creating it in another situation, so that will need more experimentation. Leaving that part of the code as it was for now, but applying the style fixes for readability.

Reply:
Just found the top style fix mentioned here, missed it when fixing the others